src/engine/TransitivePathBase.h

```diff
@@ -0,0 +1,264 @@
+// Chair of Algorithms and Data Structures.
+// Author: Florian Kramer (florian.kramer@neptun.uni-freiburg.de)
+//         Johannes Herrmann (johannes.r.herrmann(at)gmail.com)
+
+#pragma once
+
+#include <functional>
+#include <memory>
+
+#include "engine/Operation.h"
+#include "engine/QueryExecutionTree.h"
+
+using TreeAndCol = std::pair<std::shared_ptr<QueryExecutionTree>, size_t>;
+struct TransitivePathSide {
+  // treeAndCol contains the QueryExecutionTree of this side and the column
+  // where the Ids of this side are located. This member only has a value if
+  // this side was bound.
+  std::optional<TreeAndCol> treeAndCol_;
+  // Column of the sub table where the Ids of this side are located
+  size_t subCol_;
+  std::variant<Id, Variable> value_;
+  // The column in the ouput table where this side Ids are written to.
+  // This member is set by the TransitivePath class
+  size_t outputCol_ = 0;
+
+  bool isVariable() const { return std::holds_alternative<Variable>(value_); };
+
+  bool isBoundVariable() const { return treeAndCol_.has_value(); };
+
+  std::string getCacheKey() const {
+    std::ostringstream os;
+    if (!isVariable()) {
+      os << "Id: " << std::get<Id>(value_);
+    }
+
+    os << ", subColumn: " << subCol_ << "to " << outputCol_;
+
+    if (treeAndCol_.has_value()) {
+      const auto& [tree, col] = treeAndCol_.value();
+      os << ", Subtree:\n";
+      os << tree->getCacheKey() << "with join column " << col << "\n";
+    }
+    return std::move(os).str();
+  }
+
+  bool isSortedOnInputCol() const {
+    if (!treeAndCol_.has_value()) {
+      return false;
+    }
+
+    auto [tree, col] = treeAndCol_.value();
+    const std::vector<ColumnIndex>& sortedOn =
+        tree->getRootOperation()->getResultSortedOn();
+    // TODO<C++23> use std::ranges::starts_with
+    return (!sortedOn.empty() && sortedOn[0] == col);
+  }
+};
+
+// We deliberately use the `std::` variants of a hash set and hash map because
+// `absl`s types are not exception safe.
+inline constexpr auto hash = [](Id id) {
+  return std::hash<uint64_t>{}(id.getBits());
+};
+using Set = std::unordered_set<Id, decltype(hash), std::equal_to<Id>,
+                               ad_utility::AllocatorWithLimit<Id>>;
+using Map = std::unordered_map<
+    Id, Set, decltype(hash), std::equal_to<Id>,
+    ad_utility::AllocatorWithLimit<std::pair<const Id, Set>>>;
+
+class TransitivePathBase : public Operation {
+ protected:
+  std::shared_ptr<QueryExecutionTree> subtree_;
+  TransitivePathSide lhs_;
+  TransitivePathSide rhs_;
+  size_t resultWidth_ = 2;
+  size_t minDist_;
+  size_t maxDist_;
+  VariableToColumnMap variableColumns_;
+
+ public:
+  TransitivePathBase(QueryExecutionContext* qec,
+                     std::shared_ptr<QueryExecutionTree> child,
+                     TransitivePathSide leftSide, TransitivePathSide rightSide,
+                     size_t minDist, size_t maxDist);
+
+  virtual ~TransitivePathBase() {}
+
```

Comment:
I would explicitly declare the missing functions that have to be defined by the children as
`virtual dingsbums blubb(....) = 0`, S.T. it becomes more visible what sttill has to be implemented.